embedded/sql/union_row_reader.go

```diff
@@ -63,9 +63,7 @@ func newUnionRowReader(rowReaders []RowReader) (*unionRowReader, error) {
 }
 
 func (ur *unionRowReader) onClose(callback func()) {
-	for _, r := range ur.rowReaders {
-		r.onClose(callback)
-	}
+	ur.rowReaders[0].onClose(callback)
```

Comment:
hm, I'd expect to call each callback

Reply:
It's only registering the close callback here.

And in fact we want to register this only into one child that will be closed as the last one. That's why the first child is selected and the closing of readers is done in reverse order.

This was causing issues in many places actually.

The problem is that all readers first need to be closed and only then the callback can be executed. Otherwise the close callback will be executed when readers are still active - in our the callback was closing the transaction. But while closing it, it failed on closing its associated snapshot because it still had active readers.